bin/propolis-server/src/lib/migrate/compat.rs

```diff
@@ -0,0 +1,791 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Checks for compatibility of two instance specs.
+
+use std::collections::HashMap;
+
+use crate::spec::{self, SerialPortDevice};
+
+use propolis_api_types::instance_spec::{
+    components::{
+        board::Chipset,
+        devices::{PciPciBridge, SerialPortNumber},
+    },
+    PciPath,
+};
+use thiserror::Error;
+
+trait CompatComponent {
+    fn is_compatible_with(
+        &self,
+        other: &Self,
+    ) -> Result<(), ComponentIncompatibility>;
+}
+
+trait CompatCollection {
+    fn is_compatible_with(
+        &self,
+        other: &Self,
+    ) -> Result<(), CollectionIncompatibility>;
+}
+
+#[derive(Debug, Error)]
+pub enum CompatibilityError {
+    #[error("specs have incompatible boards")]
+    Board(#[from] BoardIncompatibility),
+
+    #[error("specs have incompatible pvpanic settings")]
+    Pvpanic(#[from] PvpanicIncompatibility),
```

Comment:
take it or leave it: it seems to me that these _could_ just be `#[error(transparent)]`; I don't think "specs have incompatible boards" and then hiding the actual problem in the error's cause chain provides any more useful context over the inner `BoardIncompatibility` error's message?